sgkit/tests/test_association.py

```diff
@@ -0,0 +1,169 @@
+    b_zero_slice : slice
+        Variant beta values to zero out, defaults to `slice(m // 2)`
+        meaning that the first half will all be 0.
+        Set to `slice(0)` to disable.
+
+    Returns
+    -------
+    g : (n, m) array-like
+        Simulated genotype dosage
+    x : (n, p) array-like
+        Simulated covariates
+    bg : (m,) array-like
+        Variant betas
+    ys : (m, n) array-like
+        Outcomes for each column in genotypes i.e. variant
+    """
+    if b_zero_slice is None:
+        b_zero_slice = slice(m // 2)
+    rs = np.random.RandomState(seed)
+    g = rs.uniform(size=(n, m), low=0, high=2)
+    x = rs.normal(size=(n, p))
+    bg = rs.normal(size=m)
+    bg[b_zero_slice or slice(m // 2)] = 0
+    bx = rs.normal(size=p)
+    e = rs.normal(size=n, scale=e_std)
+
+    # Simulate y values using each variant independently
+    ys = np.array([g[:, i] * bg[i] + x @ bx + e for i in range(m)])
+    return g, x, bg, ys
+
+
+def _generate_test_dataset(**kwargs: Any) -> Dataset:
+    g, x, bg, ys = _generate_test_data(**kwargs)
+    data_vars = {}
+    data_vars["dosage"] = (["variant", "sample"], g.T)
+    for i in range(x.shape[1]):
+        data_vars[f"covar_{i}"] = (["sample"], x[:, i])
+    for i in range(len(ys)):
+        data_vars[f"trait_{i}"] = (["sample"], ys[i])
+    attrs = dict(beta=bg)
+    return xr.Dataset(data_vars, attrs=attrs)  # type: ignore[arg-type]
+
+
+@pytest.fixture  # type: ignore[misc]
+def ds() -> Dataset:
+    return _generate_test_dataset()
+
+
+def _sm_statistics(
+    ds: Dataset, i: int, add_intercept: bool
+) -> RegressionResultsWrapper:
+    X = []
+    # Make sure first independent variable is variant
+    X.append(ds["dosage"].values[i])
+    for v in [c for c in list(ds.keys()) if c.startswith("covar_")]:
+        X.append(ds[v].values)
+    if add_intercept:
+        X.append(np.ones(ds.dims["sample"]))
+    X = np.stack(X).T
+    y = ds[f"trait_{i}"].values
+
+    return sm.OLS(y, X, hasconst=True).fit()
+
+
+def _get_statistics(
+    ds: Dataset, add_intercept: bool, **kwargs: Any
+) -> Tuple[DataFrame, DataFrame]:
+    df_pred: List[Dict[str, Any]] = []
+    df_true: List[Dict[str, Any]] = []
+    for i in range(ds.dims["variant"]):
+        dsr = gwas_linear_regression(
+            ds,
+            dosage="dosage",
+            trait=f"trait_{i}",
+            add_intercept=add_intercept,
+            **kwargs,
+        )
+        res = _sm_statistics(ds, i, add_intercept)
+        df_pred.append(
+            dsr.to_dataframe()  # type: ignore[no-untyped-call]
+            .rename(columns=lambda c: c.replace("variant/", ""))
+            .iloc[i]
+            .to_dict()
+        )
+        df_true.append(dict(t_value=res.tvalues[0], p_value=res.pvalues[0]))
+    return pd.DataFrame(df_pred), pd.DataFrame(df_true)
+
+
+def test_linear_regression_statistics(ds):
+    def validate(dfp: DataFrame, dft: DataFrame) -> None:
+        print(dfp)
```

Comment:
Are these `print` intentional?

Reply:
I was planning to do that often with these method tests since they'll only be shown when there's an error, and I don't often find smart diffs from assertion utils (e.g. np.testing) to be as good for debugging as simply seeing the datasets when they're small.  I take it as a somewhat acceptable convention since I've seen it in Xarray/Dask tests, and haven't seen any downsides other than that stdout is a mess if not captured by pytest (can't think of why that would ever be necessary though).